### PR TITLE
Feat/upgraded mul

### PR DESCRIPTION
**Pull request by jtlap**
il faut bencher

### CODE REVIEW

File: include/eve/module/real/core/function/promoted/simd/x86/mul.hpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+#include <eve/detail/apply_over.hpp>
+#include <eve/pattern.hpp>
+#include <concepts>
+#include <bit>
+
+namespace eve::detail
+{
+  template<unsigned_scalar_value T, typename N>
+  EVE_FORCEINLINE
+  auto mul_(EVE_SUPPORTS(sse2_), promoted_type const &, wide<T, N> a, wide<T, N> b) noexcept
```
Comment (jfalcou) [inline include/eve/module/real/core/function/promoted/simd/x86/mul.hpp:33]: as the tests shows, the intrinsics used here are SSE4+, so maybe we can just `EVE_SUPPORTS(sse4_1)` ?

File: include/eve/module/real/core/function/promoted/simd/x86/mul.hpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    using in_t = wide<T, N>;
+    using ou_t = wide<upgrade_t<T>, N>;
+    constexpr auto c = categorize<in_t>();
+    if constexpr(c == category::int32x16 && current_api >= avx512)
+    {
+      using tmp_t = wide < upgrade_t<T>, fixed<8>>;
+      constexpr auto slo = eve::fix_pattern<16>([](auto i, auto) { return i&1 ?  -1: i/2; });
+      constexpr auto shi = eve::fix_pattern<16>([](auto i, auto) { return i&1 ?  -1: 8+i/2; });
+
+      tmp_t alo = std::bit_cast<tmp_t>(a[slo]);
```
Comment (jfalcou) [inline include/eve/module/real/core/function/promoted/simd/x86/mul.hpp:154]: I'm flagging this line but : why not using eve::bitcast ?

File: test/unit/module/real/core/mul.cpp [hunk 3]
```diff
@@ -11,135 +11,181 @@
...
+//         , eve::test::generate ( eve::test::randoms(mini, 127)
+//                               , eve::test::randoms(mini, 127)
+//                               , eve::test::randoms(mini, 127)
+//                               )
+//         )
+// <typename T>( T const& a0, T const& a1, T const& a2)
+// {
+//   using eve::mul;
+//   using eve::saturated;
+//   using eve::detail::map;
```
Comment (jfalcou) [inline test/unit/module/real/core/mul.cpp:151]: why all the comments ?

File: test/unit/module/real/core/mul.cpp [hunk 4]
```diff
@@ -11,135 +11,181 @@
...
+    using pv_t = eve::detail::upgrade_t<v_t>;
+//     auto a = eve::promoted(mul)(a0, a1);
+//     std::cout << a << std::endl;
+//     TTS_EQUAL(0, 0);
+//    TTS_EQUAL( eve::promoted(mul)(a0, a2), map([](auto e, auto f) { return mul(pv_t(e), pv_t(f)); }, a0, a2));
+    TTS_EQUAL( eve::promoted(mul)(a0, a1), map([](auto e, auto f) { return mul(pv_t(e), pv_t(f)); }, a0, a1));
+//   constexpr int S = 8;
+//   eve::wide < uint32_t, eve::fixed<S>> a(eve::valmax(eve::as<uint32_t>()));
+//   eve::wide < uint32_t, eve::fixed<S>> b{[](auto i, auto){return i; }};
+//   std::cout << std::hex << " a "<< a << std::endl;
```
Comment (jfalcou) [inline test/unit/module/real/core/mul.cpp:182]: remember to cleanup

**Comment by jtlap** [comment]
must wait for interleave for performance reasons, to rebench after

**Comment by jtlap** [comment]
joel peux-tu regarder pour arm je ne pige pas ce qui se passe et je ne peux pas compiler sur place

**Pull request closed without merging**